### PR TITLE
Add overloads to child_process.spawn for when options.stdio is a tuple of 3

### PR DESCRIPTION
**Pull request by KSXGitHub**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [child_process options.stdio](https://nodejs.org/api/child_process.html#child_process_options_stdio)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by KSXGitHub** [comment]
It requires TypeScript 2.8+, how do I avoid breaking changes?

**Comment by Flarna** [comment]
The current only way to add something requiring a new typescript version is to type this twice. Once in subfolder ts3.2 for typescript >=3.2 and once in another file for < 3.2. The common part should stay in e.g. child_process.d.ts.
This mechanism results in quite a maintenance pain and as a result it's not used a lot.

**Comment by KSXGitHub** [comment]
I have worked around old TypeScript issue by using a bunch of overloads instead of ternary operators.

**Comment by KSXGitHub** [comment]
I can reduce the number of overloads by defining a type/interface for `spawn`, but I want to do it after this pull request is resolved.

File: types/node/child_process.d.ts [hunk 1]
```diff
@@ -128,9 +146,99 @@ declare module "child_process" {
         stdio?: 'pipe' | Array<null | undefined | 'pipe'>;
     }
 
+    type StdioNull = 'inherit' | 'ignore' | Stream;
+    type StdioPipe = undefined | null | 'pipe';
+
+    interface SpawnOptionsWithStdioTuple<
+        Stdin extends StdioNull | StdioPipe,
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply types/node/child_process.d.ts]: Yes, but I do not know whether `'ipc'` will make `stdin`, `stdout` and `stderr` a `null` or not. This pull requests only add more overloads in addition to existing ones, which should already allow `'ipc'` (`stdin`, `stdout` and `stderr` in this case will be `Stream | null`).

**Comment by Flarna** [comment]
I agree with @SimonSchick that child_process and crypto are meanwhile hard to review as that much overload exists. I'm no poweruser of both modules therefore I have no environment to verify.

The added complication is that we have to support TS 2.0 and can't use new TS features which would ease at least some of these cases.

Anyhow, for this PR it seems that it doesn't break any existing tests and it adds a lot new scenarios so I assume it's valid and therefore I will approve. If it breaks something it's just an indication that tests were missing and this needs to be fixed once it happens.

**Review by Flarna — APPROVED** [review]

**Pull request merged**